File: ios/ContractRunner/Sources/ContractRunner/UserMutations.swift
import Foundation

/// Pure, deterministic model for optimistic user-mutation state.
/// Mirrors `android/core-domain/.../optimistic/UserMutations.kt` 1:1.

public enum MutationKind: String, Codable, Equatable {
    case watchlist
    case titleWatched = "title_watched"
    case episodeWatched = "episode_watched"
    case seasonWatched = "season_watched"
    case rating
}

/// Pure copy of the content-type taxonomy the watch-history backend needs.
/// Mirrors `android/core-domain/.../optimistic/UserMutations.kt`.
public enum MediaContentType: String, Codable, Equatable {
    case movie
    case series
    case anime
}

public enum MutationStatus: Equatable {
    case pending
    case inflight
    case failed(reason: String, retryable: Bool)
    case conflict(serverValue: String?)
}

public protocol UserMutation: Equatable {
    var id: String { get }
    var titleItemId: String { get }
    var entityId: String { get }
    var kind: MutationKind { get }
    var createdAtMs: Int64 { get }
    var attempt: Int { get }
    var status: MutationStatus { get }
    var nextAttemptAtMs: Int64 { get }
}

public struct WatchlistMutation: UserMutation {
    public let id: String
    public let titleItemId: String
    public let entityId: String
    public let createdAtMs: Int64
    public let attempt: Int
    public let status: MutationStatus
    public let nextAttemptAtMs: Int64
    public let desired: Bool
    public var kind: MutationKind { .watchlist }
    public init(id: String, titleItemId: String, entityId: String, createdAtMs: Int64, attempt: Int, status: MutationStatus, nextAttemptAtMs: Int64, desired: Bool) {
        self.id = id
        self.titleItemId = titleItemId
        self.entityId = entityId
        self.createdAtMs = createdAtMs
        self.attempt = attempt
        self.status = status
        self.nextAttemptAtMs = nextAttemptAtMs
        self.desired = desired
    }
}

public struct TitleWatchedMutation: UserMutation {
    public let id: String
    public let titleItemId: String
    public let entityId: String
    public let createdAtMs: Int64
    public let attempt: Int
    public let status: MutationStatus
    public let nextAttemptAtMs: Int64
    public let contentType: MediaContentType
    public let desired: Bool
    public var kind: MutationKind { .titleWatched }
    public init(id: String, titleItemId: String, entityId: String, createdAtMs: Int64, attempt: Int, status: MutationStatus, nextAttemptAtMs: Int64, contentType: MediaContentType = .movie, desired: Bool) {
        self.id = id
        self.titleItemId = titleItemId
        self.entityId = entityId
        self.createdAtMs = createdAtMs
        self.attempt = attempt
        self.status = status
        self.nextAttemptAtMs = nextAttemptAtMs
        self.contentType = contentType
        self.desired = desired
    }
}


public struct RatingMutation: UserMutation {
    public let id: String
    public let titleItemId: String
    public let entityId: String
    public let createdAtMs: Int64
    public let attempt: Int
    public let status: MutationStatus
    public let nextAttemptAtMs: Int64
    public let desired: Int?
    public var kind: MutationKind { .rating }
    public init(id: String, titleItemId: String, entityId: String, createdAtMs: Int64, attempt: Int, status: MutationStatus, nextAttemptAtMs: Int64, desired: Int?) {
        self.id = id
        self.titleItemId = titleItemId
        self.entityId = entityId
        self.createdAtMs = createdAtMs
        self.attempt = attempt
        self.status = status
        self.nextAttemptAtMs = nextAttemptAtMs
        self.desired = desired
    }
}

public struct EpisodeWatchedMutation: UserMutation {
    public let id: String
    public let titleItemId: String
    public let entityId: String
    public let createdAtMs: Int64
    public let attempt: Int
    public let status: MutationStatus
    public let nextAttemptAtMs: Int64
    public let itemId: String
    public let season: Int
    public let episode: Int?
    public let videoId: String
    public let desired: Bool
    public var kind: MutationKind { .episodeWatched }
    public init(id: String, titleItemId: String, entityId: String, createdAtMs: Int64, attempt: Int, status: MutationStatus, nextAttemptAtMs: Int64, itemId: String, season: Int, episode: Int?, videoId: String, desired: Bool) {
        self.id = id
        self.titleItemId = titleItemId
        self.entityId = entityId
        self.createdAtMs = createdAtMs
        self.attempt = attempt
        self.status = status
        self.nextAttemptAtMs = nextAttemptAtMs
        self.itemId = itemId
        self.season = season
        self.episode = episode
        self.videoId = videoId
        self.desired = desired
    }
}

public struct SeasonWatchedMutation: UserMutation {
    public let id: String
    public let titleItemId: String
    public let entityId: String
    public let createdAtMs: Int64
    public let attempt: Int
    public let status: MutationStatus
    public let nextAttemptAtMs: Int64
    public let seasonItemId: String
    public let seasonNumber: Int
    public let desired: Bool
    public var kind: MutationKind { .seasonWatched }
    public init(id: String, titleItemId: String, entityId: String, createdAtMs: Int64, attempt: Int, status: MutationStatus, nextAttemptAtMs: Int64, seasonItemId: String, seasonNumber: Int, desired: Bool) {
        self.id = id
        self.titleItemId = titleItemId
        self.entityId = entityId
        self.createdAtMs = createdAtMs
        self.attempt = attempt
        self.status = status
        self.nextAttemptAtMs = nextAttemptAtMs
        self.seasonItemId = seasonItemId
        self.seasonNumber = seasonNumber
        self.desired = desired
    }
}

public struct UserStateSnapshot {
    public let isInWatchlist: Bool
    public let isWatched: Bool
    public let isRated: Bool
    public let userRating: Int?
    public let episodeWatched: [String: Bool]
    public let seasonWatched: [Int: Bool]
    public init(isInWatchlist: Bool = false, isWatched: Bool = false, isRated: Bool = false, userRating: Int? = nil, episodeWatched: [String: Bool] = [:], seasonWatched: [Int: Bool] = [:]) {
        self.isInWatchlist = isInWatchlist
        self.isWatched = isWatched
        self.isRated = isRated
        self.userRating = userRating
        self.episodeWatched = episodeWatched
        self.seasonWatched = seasonWatched
    }
}

public enum FieldSync: String, Equatable {
    case idle
    case syncing
    case error
}

public struct MutationSyncView: Equatable {
    public let status: FieldSync
    public let errorMessage: String?
    public init(status: FieldSync, errorMessage: String? = nil) {
        self.status = status
        self.errorMessage = errorMessage
    }
}

public struct DerivedUserState {
    public let watchlist: (Bool, MutationSyncView)
    public let titleWatched: (Bool, MutationSyncView)
    public let rating: (Int?, MutationSyncView)
    public let episodeWatched: [String: (Bool, MutationSyncView)]
    public let seasonWatched: [Int: (Bool, MutationSyncView)]
}

public struct RetryPolicy {
    public let baseDelayMs: Int64
    public let maxDelayMs: Int64
    public let maxAttempts: Int
    public init(baseDelayMs: Int64 = 1000, maxDelayMs: Int64 = 60000, maxAttempts: Int = 8) {
        self.baseDelayMs = baseDelayMs
        self.maxDelayMs = maxDelayMs
        self.maxAttempts = maxAttempts
    }
}

public struct OutboxAction: Equatable {
    public let mutationId: String
    public let kind: MutationKind
    public init(mutationId: String, kind: MutationKind) {
        self.mutationId = mutationId
        self.kind = kind
    }
}

public func nextBackoffDelayMs(attempt: Int, policy: RetryPolicy) -> Int64 {
    if attempt <= 0 { return policy.baseDelayMs }
    let exponent = min(attempt - 1, 30)
    let shifted: Int64
    if exponent >= 63 {
        shifted = Int64.max
    } else {
        shifted = policy.baseDelayMs * (1 << exponent)
    }
    return min(max(shifted, 0), policy.maxDelayMs)
}

public func coalesceMutations(_ mutations: [any UserMutation]) -> [any UserMutation] {
    if mutations.isEmpty { return [] }
    var byKey: [String: any UserMutation] = [:]
    for mutation in mutations {
        let key = mutation.kind.rawValue + "#" + mutation.entityId
        if let existing = byKey[key] {
            if mutation.createdAtMs >= existing.createdAtMs {
                byKey[key] = mutation
            }
        } else {
            byKey[key] = mutation
        }
    }
    return byKey.values.sorted { ($0.createdAtMs, $0.id) < ($1.createdAtMs, $1.id) }
}

private func latestOf(_ group: [any UserMutation]) -> any UserMutation? {
    group.max { ($0.createdAtMs, $0.id) < ($1.createdAtMs, $1.id) }
}

public func deriveUserState(snapshot: UserStateSnapshot, mutations: [any UserMutation]) -> DerivedUserState {
    let grouped = Dictionary(grouping: mutations, by: { $0.kind })

    func active(_ kind: MutationKind) -> any UserMutation? {
        latestOf(grouped[kind] ?? [])
    }

    func reduce<T>(server: T, mutation: (any UserMutation)?, desiredOf: (any UserMutation) -> T) -> (T, MutationSyncView) {
        guard let mutation = mutation else {
            return (server, MutationSyncView(status: .idle))
        }
        switch mutation.status {
        case .pending, .inflight:
            return (desiredOf(mutation), MutationSyncView(status: .syncing))
        case let .failed(reason, _):
            let message = reason.isEmpty ? "Update failed." : reason
            return (server, MutationSyncView(status: .error, errorMessage: message))
        case let .conflict(serverValue):
            let message = serverValue.map { "Server has a different value: \($0)" } ?? "Server rejected the change."
            return (server, MutationSyncView(status: .error, errorMessage: message))
        }
    }

    let watchlist = reduce(server: snapshot.isInWatchlist, mutation: active(.watchlist)) { ($0 as! WatchlistMutation).desired }
    let titleWatched = reduce(server: snapshot.isWatched, mutation: active(.titleWatched)) { ($0 as! TitleWatchedMutation).desired }
    let rating = reduce(server: snapshot.userRating, mutation: active(.rating)) { ($0 as! RatingMutation).desired }

    var episodeWatched = snapshot.episodeWatched
    var episodeSync: [String: MutationSyncView] = [:]
    for mutation in grouped[.episodeWatched] ?? [] {
        let m = mutation as! EpisodeWatchedMutation
        let derived = reduce(server: episodeWatched[m.videoId] ?? false, mutation: m) { ($0 as! EpisodeWatchedMutation).desired }
        episodeWatched[m.videoId] = derived.0
        episodeSync[m.videoId] = derived.1
    }

    var seasonWatched = snapshot.seasonWatched
    var seasonSync: [Int: MutationSyncView] = [:]
    for mutation in grouped[.seasonWatched] ?? [] {
        let m = mutation as! SeasonWatchedMutation
        let derived = reduce(server: seasonWatched[m.seasonNumber] ?? false, mutation: m) { ($0 as! SeasonWatchedMutation).desired }
        seasonWatched[m.seasonNumber] = derived.0
        seasonSync[m.seasonNumber] = derived.1
    }

    let episodeDerived = Dictionary(
        uniqueKeysWithValues: episodeWatched.map { (key, value) in
            (key, (value, episodeSync[key] ?? MutationSyncView(status: .idle)))
        }
    )
    let seasonDerived = Dictionary(
        uniqueKeysWithValues: seasonWatched.map { (key, value) in
            (key, (value, seasonSync[key] ?? MutationSyncView(status: .idle)))
        }
    )

    return DerivedUserState(
        watchlist: watchlist,
        titleWatched: titleWatched,
        rating: rating,
        episodeWatched: episodeDerived,
        seasonWatched: seasonDerived
    )
}

public func planOutbox(mutations: [any UserMutation], nowMs: Int64) -> [OutboxAction] {
    mutations
        .filter { mutation in
            if case .pending = mutation.status { return mutation.nextAttemptAtMs <= nowMs }
            return false
        }
        .sorted { ($0.createdAtMs, $0.id) < ($1.createdAtMs, $1.id) }
        .map { OutboxAction(mutationId: $0.id, kind: $0.kind) }
}

File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/optimistic/UserMutations.kt
package com.crispy.tv.domain.optimistic

/**
 * Pure, deterministic model for optimistic user-mutation state.
 *
 * Local intent (a [UserMutation]) is applied on top of the last known server
 * truth ([UserStateSnapshot]) via [deriveUserState]. The in-flight outbox is
 * scheduled by [planOutbox]. Nothing here touches IO, Android types, or the
 * system clock, so it can be mirrored 1:1 by the Swift ContractRunner and
 * covered by contract fixtures.
 */

enum class MutationKind {
    WATCHLIST,
    TITLE_WATCHED,
    EPISODE_WATCHED,
    SEASON_WATCHED,
    RATING,
}

/**
 * Pure copy of the content-type taxonomy the watch-history backend needs, kept
 * in [core-domain] so mutations stay free of `player`/Android types. The app
 * maps to/from [com.crispy.tv.player.MetadataLabMediaType] at the boundary.
 */
enum class MediaContentType {
    MOVIE,
    SERIES,
    ANIME,
}

/**
 * Lifecycle of a single mutation. [Inflight] is transient (never persisted);
 * on reload it is coerced back to [Pending] so a crashed write is retried.
 * A successfully synced mutation is removed entirely from the store.
 */
sealed interface MutationStatus {
    data object Pending : MutationStatus

    data object Inflight : MutationStatus

    data class Failed(
        val reason: String,
        val retryable: Boolean,
    ) : MutationStatus

    data class Conflict(
        val serverValue: String?,
    ) : MutationStatus
}

/**
 * A single user intent. `id` is a client-generated nonce used for idempotency
 * on the server and as the dedupe key for retries. `entityId` groups mutations
 * that target the same logical target so rapid toggles coalesce; `titleItemId`
 * lets the UI observe every pending mutation for one title at once.
 */
sealed interface UserMutation {
    val id: String
    val titleItemId: String
    val entityId: String
    val kind: MutationKind
    val createdAtMs: Long
    val attempt: Int
    val status: MutationStatus
    val nextAttemptAtMs: Long
}

data class WatchlistMutation(
    override val id: String,
    override val titleItemId: String,
    override val entityId: String,
    override val createdAtMs: Long,
    override val attempt: Int,
    override val status: MutationStatus,
    override val nextAttemptAtMs: Long,
    val desired: Boolean,
) : UserMutation {
    override val kind: MutationKind = MutationKind.WATCHLIST
}

data class TitleWatchedMutation(
    override val id: String,
    override val titleItemId: String,
    override val entityId: String,
    override val createdAtMs: Long,
    override val attempt: Int,
    override val status: MutationStatus,
    override val nextAttemptAtMs: Long,
    val contentType: MediaContentType,
    val desired: Boolean,
) : UserMutation {
    override val kind: MutationKind = MutationKind.TITLE_WATCHED
}

data class RatingMutation(
    override val id: String,
    override val titleItemId: String,
    override val entityId: String,
    override val createdAtMs: Long,
    override val attempt: Int,
    override val status: MutationStatus,
    override val nextAttemptAtMs: Long,
    /** `null` means "remove rating". */
    val desired: Int?,
) : UserMutation {
    override val kind: MutationKind = MutationKind.RATING
}

data class EpisodeWatchedMutation(
    override val id: String,
    override val titleItemId: String,
    override val entityId: String,
    override val createdAtMs: Long,
    override val attempt: Int,
    override val status: MutationStatus,
    override val nextAttemptAtMs: Long,
    val itemId: String,
    val season: Int,
    val episode: Int?,
    val videoId: String,
    val desired: Boolean,
) : UserMutation {
    override val kind: MutationKind = MutationKind.EPISODE_WATCHED
}

data class SeasonWatchedMutation(
    override val id: String,
    override val titleItemId: String,
    override val entityId: String,
    override val createdAtMs: Long,
    override val attempt: Int,
    override val status: MutationStatus,
    override val nextAttemptAtMs: Long,
    val seasonItemId: String,
    val seasonNumber: Int,
    val desired: Boolean,
) : UserMutation {
    override val kind: MutationKind = MutationKind.SEASON_WATCHED
}

/**
 * Last known server truth for a title. Fields that have not been loaded yet
 * should be represented as absent (`false`/`null`) rather than guessed; the
 * merge only overrides values for which a pending mutation exists.
 */
data class UserStateSnapshot(
    val isInWatchlist: Boolean = false,
    val isWatched: Boolean = false,
    val isRated: Boolean = false,
    val userRating: Int? = null,
    val episodeWatched: Map<String, Boolean> = emptyMap(),
    val seasonWatched: Map<Int, Boolean> = emptyMap(),
)

enum class FieldSync {
    IDLE,
    SYNCING,
    ERROR,
}

data class MutationSyncView(
    val status: FieldSync,
    val errorMessage: String? = null,
)

data class DerivedUserState(
    /** `value` is the rating (null == unrated); `isRated` is derived from it. */
    val watchlist: Pair<Boolean, MutationSyncView>,
    val titleWatched: Pair<Boolean, MutationSyncView>,
    val rating: Pair<Int?, MutationSyncView>,
    val episodeWatched: Map<String, Pair<Boolean, MutationSyncView>>,
    val seasonWatched: Map<Int, Pair<Boolean, MutationSyncView>>,
)

data class RetryPolicy(
    val baseDelayMs: Long = 1000,
    val maxDelayMs: Long = 60_000,
    val maxAttempts: Int = 8,
)

data class OutboxAction(
    val mutationId: String,
    val kind: MutationKind,
)

/**
 * Deterministic backoff: `base * 2^(attempt-1)`, capped. Pure so the contract
 * fixture and the runtime processor agree.
 */
fun nextBackoffDelayMs(
    attempt: Int,
    policy: RetryPolicy,
): Long {
    if (attempt <= 0) return policy.baseDelayMs
    val exponent = (attempt - 1).coerceAtMost(30)
    val multiplied =
        try {
            policy.baseDelayMs * (1L shl exponent)
        } catch (_: ArithmeticException) {
            Long.MAX_VALUE
        }
    return multiplied.coerceAtMost(policy.maxDelayMs).coerceAtLeast(0)
}

/**
 * Keep at most one active mutation per (kind, entityId), choosing the most
 * recently created. Older intents for the same target are superseded — this is
 * what makes rapid double-toggles collapse into a single final write.
 */
fun coalesceMutations(mutations: List<UserMutation>): List<UserMutation> {
    if (mutations.isEmpty()) return emptyList()
    val byKey = mutableMapOf<String, UserMutation>()
    for (mutation in mutations) {
        val key = mutation.kind.name + "#" + mutation.entityId
        val existing = byKey[key]
        if (existing == null || mutation.createdAtMs >= existing.createdAtMs) {
            byKey[key] = mutation
        }
    }
    return byKey.values.sortedWith(compareBy({ it.createdAtMs }, { it.id }))
}

private fun latestOf(
    group: List<UserMutation>,
): UserMutation? =
    group.maxWithOrNull(compareBy({ it.createdAtMs }, { it.id }))

/**
 * Apply pending/local intent on top of server truth.
 *
 * Rules (deterministic, latest-created wins within a kind+entity):
 * - [MutationStatus.Pending] / [MutationStatus.Inflight]: the local `desired`
 *   value is shown and the field is marked [FieldSync.SYNCING].
 * - [MutationStatus.Failed] / [MutationStatus.Conflict]: fall back to server
 *   truth and mark [FieldSync.ERROR] so the UI can offer a retry.
 * - No active mutation: server truth, [FieldSync.IDLE].
 */
fun deriveUserState(
    snapshot: UserStateSnapshot,
    mutations: List<UserMutation>,
): DerivedUserState {
    val grouped = mutations.groupBy { it.kind }

    fun active(kind: MutationKind): UserMutation? = latestOf(grouped[kind].orEmpty())

    fun <T> reduce(
        server: T,
        mutation: UserMutation?,
        desiredOf: (UserMutation) -> T,
    ): Pair<T, MutationSyncView> {
        if (mutation == null) {
            return server to MutationSyncView(FieldSync.IDLE)
        }
        return when (val status = mutation.status) {
            MutationStatus.Pending,
            MutationStatus.Inflight,
            -> desiredOf(mutation) to MutationSyncView(FieldSync.SYNCING)

            is MutationStatus.Failed ->
                server to
                    MutationSyncView(
                        FieldSync.ERROR,
                        status.reason.takeIf { it.isNotBlank() } ?: "Update failed.",
                    )

            is MutationStatus.Conflict ->
                server to
                    MutationSyncView(
                        FieldSync.ERROR,
                        status.serverValue?.let { "Server has a different value: $it" }
                            ?: "Server rejected the change.",
                    )
        }
    }

    val watchlist = reduce(snapshot.isInWatchlist, active(MutationKind.WATCHLIST)) { (it as WatchlistMutation).desired }
    val titleWatched = reduce(snapshot.isWatched, active(MutationKind.TITLE_WATCHED)) { (it as TitleWatchedMutation).desired }
    val rating = reduce(snapshot.userRating, active(MutationKind.RATING)) { (it as RatingMutation).desired }

    val episodeWatched = snapshot.episodeWatched.toMutableMap()
    val episodeSync = mutableMapOf<String, MutationSyncView>()
    for (mutation in grouped[MutationKind.EPISODE_WATCHED].orEmpty()) {
        mutation as EpisodeWatchedMutation
        val derived = reduce(episodeWatched[mutation.videoId] ?: false, mutation) { mutation.desired }
        episodeWatched[mutation.videoId] = derived.first
        episodeSync[mutation.videoId] = derived.second
    }

    val seasonWatched = snapshot.seasonWatched.toMutableMap()
    val seasonSync = mutableMapOf<Int, MutationSyncView>()
    for (mutation in grouped[MutationKind.SEASON_WATCHED].orEmpty()) {
        mutation as SeasonWatchedMutation
        val derived = reduce(seasonWatched[mutation.seasonNumber] ?: false, mutation) { mutation.desired }
        seasonWatched[mutation.seasonNumber] = derived.first
        seasonSync[mutation.seasonNumber] = derived.second
    }

    return DerivedUserState(
        watchlist = watchlist,
        titleWatched = titleWatched,
        rating = rating,
        episodeWatched =
            episodeWatched
                .mapValues { (key, value) ->
                    value to (episodeSync[key] ?: MutationSyncView(FieldSync.IDLE))
                },
        seasonWatched =
            seasonWatched
                .mapValues { (key, value) ->
                    value to (seasonSync[key] ?: MutationSyncView(FieldSync.IDLE))
                },
    )
}

/**
 * Select due mutations to execute now. Only [MutationStatus.Pending] entries
 * whose `nextAttemptAtMs` has passed are eligible; results are ordered
 * deterministically so the processor and contract fixture agree.
 */
fun planOutbox(
    mutations: List<UserMutation>,
    nowMs: Long,
): List<OutboxAction> =
    mutations
        .filter { it.status == MutationStatus.Pending && it.nextAttemptAtMs <= nowMs }
        .sortedWith(compareBy({ it.createdAtMs }, { it.id }))
        .map { OutboxAction(mutationId = it.id, kind = it.kind) }
